src/tests/hotreload-integration-test.ts

```diff
@@ -0,0 +1,164 @@
+        return class extends DomParticle {
+          get template() { return 'Hello <span>{{name}}</span>, old age: <span>{{age}}</span>'; }
+
+          render() {
+            return {name: 'Jack', age: '10'};
+          }
+        };
+      });`
+    });
+
+    const newLoader = new StubLoader({
+      'A.js': `defineParticle(({DomParticle}) => {
+        return class extends DomParticle {
+          get template() { return 'Hello <span>{{name}}</span>, new age: <span>{{age}}</span>'; }
+
+          render() {
+            return {name: 'Jack', age: '15'};
+          }
+        };
+      });`
+    });
+
+    const id = ArcId.newForTest('demo');
+    const pecFactories = [StubPecFactory(loader).bind(null)];
+    const storageKey = 'volatile://' + id.toString();
+    const slotComposer = new StubSlotComposer();
+    
+    const arc = new Arc({id, pecFactories, storageKey, slotComposer, loader, context});
+
+    const [recipe] = arc.context.recipes;
+    assert.isTrue(recipe.normalize() && recipe.isResolved());
+    await arc.instantiate(recipe);
+
+    const slotConsumer = slotComposer.consumers[0] as HeadlessSlotDomConsumer;
+
+    const oldModel = slotConsumer.getRendering().model;
+    const oldTemplate = slotConsumer._content.template;
+    assert.deepEqual(oldModel,  {name: 'Jack', age: '10'});
+    assert.deepEqual(oldTemplate, `Hello <span>{{name}}</span>, old age: <span>{{age}}</span>`);
+
+    Object.assign(arc._loader, newLoader);
+    arc.pec.reload(arc.pec.particles);
+    await arc.idle;
+
+    const newModel = slotConsumer.getRendering().model;
+    const newTemplate = slotConsumer._content.template;
+
+    assert.deepEqual(newModel,  {name: 'Jack', age: '15'});
+    assert.deepEqual(newTemplate, `Hello <span>{{name}}</span>, new age: <span>{{age}}</span>`);
+  });
+});
+
+describe('Hot Code Reload for WASM Particle', async () => {
+  before(function() {
+    if (!global['testFlags'].enableWasm) {
+      this.skip();
+    }
+  });
+
+  it('updates model and template', async () => {
+    const loader = new StubWasmLoader();
+    const manifestString = `
+      particle HotReloadTest in 'build/tests/source/test-module-old.wasm'
+        consume root
+
+      recipe
+        slot 'rootslotid-root' as slot0
+        HotReloadTest
+          consume root as slot0`;
+
+    const context = await Manifest.parse(manifestString);
+    const id = ArcId.newForTest('demo');
+    const pecFactories = [StubPecFactory(loader).bind(null)];
+    const storageKey = 'volatile://' + id.toString();
+    const slotComposer = new RozSlotComposer();
+    
+    const arc = new Arc({id, pecFactories, storageKey, slotComposer, loader, context});
+  
+    const [recipe] = arc.context.recipes;
+    assert.isTrue(recipe.normalize() && recipe.isResolved());
+    await arc.instantiate(recipe);
+    await arc.idle;
+
+    loader.reloaded = true;
+    arc.pec.reload(arc.pec.particles);
+    await arc.idle;
+
+    assert.deepStrictEqual(slotComposer.received, [
```

Comment:
Any reason why you do assertions with slot composer differently in the JS and WASM tests?

Reply:
There wasn't any particular reasons.. I updated the test so they look similar to each other! :D

Reply:
Closing the loop: the difference was due to different SlotComposer impls being used. The JS test originally used a local override class, while wasm used RozSlotComposer.

FWIW I think both should use the latter - observing simple rendering interactions (as opposed to setting up more complex expectations) is the reason I added RozSlotComposer.